jupyter_collaboration/loaders.py

```diff
@@ -0,0 +1,94 @@
+import asyncio
+from logging import Logger, getLogger
+from typing import Any, Callable, Coroutine, Dict, Optional
+
+from jupyter_server.utils import ensure_async
+
+
+class FileLoader:
+    def __init__(
+        self,
+        path: str,
+        file_format: str,
+        file_type: str,
+        contents_manager: Any,
+        log: Optional[Logger],
+        poll_interval: Optional[float] = None,
+    ) -> None:
+        self._path: str = path
+        self._file_format: str = file_format
+        self._file_type: str = file_type
+        self._last_modified = None
+
+        self._lock = asyncio.Lock()
+        self._poll_interval = poll_interval
+        self._contents_manager = contents_manager
+
+        self._log = log or getLogger(__name__)
+        self._subscriptions: Dict[str, Callable[[str], Coroutine[Any, Any, None]]] = {}
+
+        if self._poll_interval:
+            self._watcher = asyncio.create_task(self.watch_file())
+
+    @property
+    def path(self):
+        return self._path
+
+    def clean(self) -> None:
+        self._watcher.cancel()
+
+    def rename_file(self, path: str) -> None:
+        self._path = path
+
+    def number_of_subscriptions(self) -> int:
+        return len(self._subscriptions)
+
+    def observe(self, id: str, callback: Callable[[str], Coroutine[Any, Any, None]]) -> None:
+        self._subscriptions[id] = callback
+
+    def unobserve(self, id: str) -> None:
+        del self._subscriptions[id]
+
+    async def load_content(self, format: str, file_type: str, content: bool) -> Dict[str, Any]:
+        return await ensure_async(
+            self._contents_manager.get(self._path, format=format, type=file_type, content=content)
+        )
+
+    async def save_content(self, model: Dict[str, Any]) -> None:
+        async with self._lock:
+            m = await self.load_content(model["format"], model["type"], False)
+
+            if self._last_modified is None or self._last_modified >= m["last_modified"]:
```

Comment:
I don't like this condition: `self._last_modified >= m["last_modified"]`. It is confusing because `self._last_modified` cannot be greater than `m["last_modified"]`, and the way it is written now lets people think that it can.
You can either rewrite it like this:
```py
if self._last_modified is None or self._last_modified == m["last_modified"]:
```
Or like that (switching the blocks):
```py
if self._last_modified is not None and self._last_modified < m["last_modified"]:
```